plugins/image-prioritizer/detect.js

```diff
@@ -0,0 +1,227 @@
+}
+
+/**
+ * Initializes extension.
+ *
+ * @since n.e.x.t
+ *
+ * @type {InitializeCallback}
+ * @param {InitializeArgs} args Args.
+ */
+export async function initialize( { isDebug, webVitalsLibrarySrc } ) {
+	const { onLCP } = await import( webVitalsLibrarySrc );
+	onLCP(
+		( /** @type {LCPMetric} */ metric ) => {
+			handleLCPMetric( metric, isDebug );
+		},
+		{
+			// This avoids needing to click to finalize LCP candidate. While this is helpful for testing, it also
+			// ensures that we always get an LCP candidate reported. Otherwise, the callback may never fire if the
+			// user never does a click or keydown, per <https://github.com/GoogleChrome/web-vitals/blob/07f6f96/src/onLCP.ts#L99-L107>.
+			reportAllChanges: true,
+		}
+	);
+}
+
+/**
+ * Gets the performance resource entry for a given URL.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} url - Resource URL.
+ * @return {PerformanceResourceTiming|null} Resource entry or null.
+ */
+function getPerformanceResourceByURL( url ) {
+	const entries =
+		/** @type PerformanceResourceTiming[] */ performance.getEntriesByType(
+			'resource'
+		);
+	for ( const entry of entries ) {
+		if ( entry.name === url ) {
+			return entry;
+		}
+	}
+	return null;
+}
+
+/**
+ * Handles a new LCP metric being reported.
+ *
+ * @since n.e.x.t
+ *
+ * @param {LCPMetric} metric  - LCP Metric.
+ * @param {boolean}   isDebug - Whether in debug mode.
+ */
+function handleLCPMetric( metric, isDebug ) {
+	for ( const entry of metric.entries ) {
+		// Look only for LCP entries that have a URL and a corresponding element which is not an IMG or VIDEO.
+		if (
+			! entry.url ||
+			! ( entry.element instanceof HTMLElement ) ||
+			entry.element instanceof HTMLImageElement ||
+			entry.element instanceof HTMLVideoElement
+		) {
+			continue;
+		}
+
+		// Always ignore data: URLs.
+		if ( entry.url.startsWith( 'data:' ) ) {
+			continue;
+		}
+
+		// Skip elements that have the background image defined inline.
+		// These are handled by Image_Prioritizer_Background_Image_Styled_Tag_Visitor.
+		if ( entry.element.style.backgroundImage ) {
+			continue;
+		}
+
+		// Now only consider proceeding with the URL if its loading was initiated with stylesheet or preload link.
+		const resourceEntry = getPerformanceResourceByURL( entry.url );
+		if (
+			! resourceEntry ||
+			! [ 'css', 'link' ].includes( resourceEntry.initiatorType )
+		) {
+			if ( isDebug ) {
+				warn(
+					`Skipped considering URL (${ entry.url }) due to unexpected performance resource timing entry:`,
+					resourceEntry
+				);
+			}
+			return;
+		}
+
+		// Skip URLs that are excessively long. This is the maxLength defined in image_prioritizer_add_element_item_schema_properties().
+		if ( entry.url.length > 500 ) {
+			if ( isDebug ) {
+				log( `Skipping very long URL: ${ entry.url }` );
+			}
+			return;
+		}
+
+		// Also skip Custom Elements which have excessively long tag names.
+		if ( entry.element.tagName.length > 25 ) {
```

Comment:
curious where these limits come from/how you decided on them.

Reply:
The longest tags in HTML are 10 characters: `FIGCAPTION` and `BLOCKQUOTE`. Oh, but in SVG the longest is `feComponentTransfer`. For Custom Element names, they could be much longer, so I recall I just asked Gemini what the longest known Custom Elements were known to be and gave this gut guess. 

However, I just checked data from Web Almanac: https://docs.google.com/spreadsheets/d/1grkd2_1xSV3jvNK6ucRQ0OL1HmGTsScHuwA8GZuRLHU/edit?gid=2057119066#gid=2057119066

There are 9 Custom Elements that have a length longer than 25:

Count | Custom Element
-- | --
26 | i-amphtml-scroll-container
26 | feefowidget-container-main
26 | mat-expansion-panel-header
26 | broadstreet-zone-container
27 | s_if_var_sidebarcategoryuse
27 | popup_notifications_wrapper
28 | mfa-finish-enrolling-message
28 | enroll-extra-factors-message
32 | more-verification-options-button

So probably should make this limit 50 and then add a note explaining where thus number comes from.

Reply:
Oh, here's [updated data](https://docs.google.com/spreadsheets/d/1TtOMr_w58HvqNBv4RIWX021Lxm6m5ajYOcRykrPdAJc/edit?gid=1606033584#gid=1606033584) [from 2024](https://almanac.httparchive.org/en/2024/markup#fig-18):

Client | Length | Custom Element | Pages
-- | -- | -- | --:
desktop | 32 | header-sidebar-collapsible-panel | 12922
mobile | 32 | header-sidebar-collapsible-panel | 14726
desktop | 29 | featured-collections-carousel | 5844
mobile | 29 | featured-collections-carousel | 6538
desktop | 28 | discourse-assets-stylesheets | 6834
mobile | 28 | discourse-assets-stylesheets | 6288
mobile | 27 | wc-order-attribution-inputs | 67920
desktop | 27 | wc-order-attribution-inputs | 50385
desktop | 26 | oem-container-bootstrapper | 5477
mobile | 26 | announcement-bar-component | 6972
desktop | 26 | announcement-bar-component | 5517

Reply:
Length increased to 100 in d2e90d36e.

Reply:
thanks for researching!